### PR TITLE
Updates from PyCharm, November 2016

### PR DESCRIPTION
**Pull request by Elizaveta239**
This pull request contains all the PyCharm debugger's changes since the previous pull request.

@fabioz We have the next PyCharm release at the end of November and we would be very grateful if you update the `pydevd` module on PyPi to make it consistent with the new version of PyCharm.
I'm very sorry for such a long pause between pull requests. This pull request contains plenty of commits, and if you have no time to review and merge all of them, please, let me know and I'll create a separate pull request which contains only the most important fixes related to api changes.
Thank you!

### CODE REVIEW

**Comment by fabioz** [comment]
I'll take a look at it, still, it seems that the test-cases are failing:

In particular, the tests from : https://travis-ci.org/fabioz/PyDev.Debugger/builds/172315632 (travis seems to be having a version info when creating the env -- if you can take a look at that it'd also be nice, otherwise I'll check how to fix that on my side).

File: _pydevd_bundle/pydevd_frame.py [hunk 1]
```diff
@@ -255,6 +259,42 @@ def handle_exception(self, frame, event, arg):
...
+                        name = get_func_name(frame)
+                        frame.f_back.f_locals[RETURN_VALUES_DICT][name] = arg
+            if main_debugger.remove_return_values_flag:
+                # Showing return values was turned off, we should remove them from locals dict.
+                # The values can be in the current frame or in the back one
+                if RETURN_VALUES_DICT in dict_keys(frame.f_locals):
+                    dict_pop(frame.f_locals, RETURN_VALUES_DICT)
+                if hasattr(frame, "f_back") and hasattr(frame.f_back, "f_locals"):
+                    if RETURN_VALUES_DICT in dict_keys(frame.f_back.f_locals):
+                        dict_pop(frame.f_back.f_locals, RETURN_VALUES_DICT)
```
Comment (fabioz) [inline _pydevd_bundle/pydevd_frame.py]: In this case, you don't need to check if it exists to remove it later, just do a dict_pop passing a default argument and remove the check to see if it's in the dict.

File: _pydevd_bundle/pydevd_frame.py [hunk 2]
```diff
@@ -255,6 +259,42 @@ def handle_exception(self, frame, event, arg):
...
+            except:
+                traceback.print_exc()
+                return func_name
+
+        try:
+            if main_debugger.show_return_values:
+                if event == "return" and hasattr(frame, "f_code") and hasattr(frame.f_code, "co_name"):
+                    if hasattr(frame, "f_back") and hasattr(frame.f_back, "f_locals"):
+                        if RETURN_VALUES_DICT not in dict_keys(frame.f_back.f_locals):
+                            frame.f_back.f_locals[RETURN_VALUES_DICT] = {}
```
Comment (fabioz) [inline _pydevd_bundle/pydevd_frame.py]: Actually, better in this case would be getting with a default None value and if it returns None, set the initial RETURN_VALUES_DICT (i.e.: no need to check and later get).

**Comment by fabioz** [comment]
Hi Elizaveta,

Given that it was a big pull request I ended up with many comments... still, I believe most should be easy to solve, so, after you're able to take a look at those I'll merge it on my side.

Also, I'd like to know if you have checked for performance regressions from the version before this commit?

**Comment by Elizaveta239** [comment]
@fabioz Thank you very much for your comments! I'll fix them as soon as possible.

What about performance, we don't check it automatically, but we usually check it manually before the releases. Is your question about performance related to `gc.get_referrers`?

**Comment by fabioz** [comment]
Not really related to that one (you'd need to have a really loaded
application to get the slowdown for that particular use-case -- which is
usually the case for real-world apps, but it'd be really hard to actually
test for it unless you're actually looking for that use-case), but related
to the overall changes...

In particular, I think that the stepping (after actually hitting a frame)
became slower if you had frames with many vars -- most should be solved by
the fixes I pointed in the review, but I think it'd still be worth actually
testing to check if some other performance regression could've happened
(so, it'd be nice to check the version before this pull request and after
it to, even if manually, to see if some performance regression slipped
there).

File: _pydevd_bundle/pydevd_utils.py [hunk 3]
```diff
@@ -120,6 +125,56 @@ def quote_smart(s, safe='/'):
         return quote(s, safe)
 
 
+def get_clsname_for_code(code, frame):
+    clsname = None
+    if GC_SUPPORTED:
+        ## use of gc.get_referrers() was suggested by Michael Hudson
+        # all functions which refer to this code object
+        funcs = [f for f in gc.get_referrers(code)
```
Comment (Elizaveta239) [inline _pydevd_bundle/pydevd_utils.py]: Yes, I like the idea to remove GC_SUPPORTED part. I've tested it and I haven't found any cases when it works better than inspecting `code.co_varnames`.

File: pydevconsole.py [hunk 4]
```diff
@@ -356,12 +359,33 @@ def start_server(host, port, client_port):
...
+
+    try:
+        if IPYTHON and hasattr(__builtin__, 'interpreter'):
+            pydev_interpreter = get_interpreter().interpreter
+            if hasattr(pydev_interpreter, 'ipython') and hasattr(pydev_interpreter.ipython, 'user_ns_hidden'):
+                try:
+                    user_hidden_dict = pydev_interpreter.ipython.user_ns_hidden
+                except:
+                    user_hidden_dict = dict([(key, val) for key, val in pydev_interpreter.ipython.user_ns.items()
+                                    if key in pydev_interpreter.ipython.user_ns_hidden])
```
Comment (Elizaveta239) [inline pydevconsole.py]: Yes, it was done in order to support Python 1.x, where `user_ns_hidden` was a set with hidden variables, not a dictionary. But, of course, you're right it's implemented incorrectly.

File: pydevd_file_utils.py [hunk 5]
```diff
@@ -73,10 +73,31 @@
...
+except ImportError:
+    CTYPES_AVAILABLE = False
+
+
+def convert_to_long_pathname(filename):
+    if CTYPES_AVAILABLE:
+        buf = ctypes.create_unicode_buffer(260)
+        GetLongPathName = ctypes.windll.kernel32.GetLongPathNameW
+        if IS_PY2:
+            filename = unicode(filename)
```
Comment (Elizaveta239) [inline pydevd_file_utils.py]: We need to pass unicode string to the function `GetLongPathNameW`, that's why we convert it.

**Comment by fabioz** [comment]
Hi Elizaveta,

I've started the process to integrate the branch, but it seems there's some critical error in Linux...

After applying the pull request, many tests on Python 2.7, in Linux, are failing with:

"OSError: [Errno 12] Cannot allocate memory"

```
ERROR: test_case_django (tests_python.test_debugger.TestPython)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/fabioz/PyDev.Debugger/tests_python/test_debugger.py", line 988, in test_case_django
    self.check_case(WriterThreadCaseDjango)
  File "/home/travis/build/fabioz/PyDev.Debugger/tests_python/debugger_unittest.py", line 174, in check_case
    return self.run_process(args, writer_thread)
  File "/home/travis/build/fabioz/PyDev.Debugger/tests_python/debugger_unittest.py", line 187, in run_process
    process = self.create_process(args, writer_thread)
  File "/home/travis/build/fabioz/PyDev.Debugger/tests_python/debugger_unittest.py", line 182, in create_process
    env=writer_thread.get_environ() if writer_thread is not None else None,
  File "/home/travis/miniconda2/lib/python2.7/subprocess.py", line 711, in __init__
    errread, errwrite)
  File "/home/travis/miniconda2/lib/python2.7/subprocess.py", line 1235, in _execute_child
    self.pid = os.fork()
OSError: [Errno 12] Cannot allocate memory
```

I've made some fixes and reran the test suite in the development branch (https://travis-ci.org/fabioz/PyDev.Debugger/builds/174557986) and all tests (both Python 2.7 and 3.5) are passing, so, all issues now seem related to the changes from this pull request (unfortunately, it's a big one, so, it may need some bisecting to discover what actually made the issue appear).

So, given that the errors now only happen after applying the pull request and are failing consistently on Travis (i.e.: Linux), can you take a look at that? As a note, I've merged the current development with your branch in https://travis-ci.org/fabioz/PyDev.Debugger/builds/174559613 -- so, you can see how it's failing consistently (initially I thought it was a Travis hiccup, but given that it's very consistent and the development version passes, this doesn't seem to be the case).

**Comment by Elizaveta239** [comment]
Hi Fabio!

Yes, I've seen this error, but I've also thought it was Travis hiccup, because I can't reproduce this error locally on my Linux machine with Python 2.7. But I'll look at this problem and try to fix it.

**Comment by Elizaveta239** [comment]
I've done a small research and measured memory usage in our tests. It turns out that the version from my pull request has  ~7.43 * 10 ^ 9 bytes in usage after the last test, and the current development version has ~7.22 * 10 ^ 9 bytes in usage (according to `psutil` tool, I ran each version 10 times on my machine locally and calculated an average value). 
Of course, the memory usage increased, but it doesn't look so critical. Travis starts throwing exceptions, when used memory becomes more than ~7,4 * 10 ^ 9.

I tried to revert some commits, but I failed to find a special one that significantly increases memory usage. How do you think is this increase really important and should I continue fixing it? Or we can just give more memory for that process on Travis agent?

**Comment by fabioz** [comment]
Well, according to https://docs.travis-ci.com/user/ci-environment/, 7.5 GB
is the limit, so, this explains it, but I'm actually surprised we're
actually getting near that limit in the tests... so, it'd be really nice if
you could investigate just a bit more to try to discover why we're actually
using all that memory in the tests... otherwise, I'll investigate it here
(but in this case, I'll only actually have time to do it towards the end of
the next week).

**Comment by Elizaveta239** [comment]
Ok, I'll try to discover the problem.

**Comment by fabioz** [comment]
Hi Elizaveta,

I ended up having some time over the weekend and already found the issue...
Just haven't pushed it to the development branch yet...

**Comment by fabioz** [comment]
Hi Elizaveta,

I've just merged this to the repo (note that I've changed the repository so that the master branch is the main branch -- the development branch was removed -- I think in the end it complicated more the workflow than it helped).

I've also done some minor fixes on top of your commits in a re-review of the code, so, you should probably sync back -- Also, all tests should be passing now -- https://travis-ci.org/fabioz/PyDev.Debugger/builds/176118419 ;)

Do you have something else you'd like to add to the debugger before a new pypi release?

**Comment by Elizaveta239** [comment]
Hi Fabio,

I'm glad to hear that! Thank you so much! :)
Yes, we've got some user's feedback and I would like to report some small fixes related to Python Console and IPython. I'll create a separate pull request.

**Comment by Elizaveta239** [comment]
Hi Fabio,

I'm glad to hear that! Thank you so much! :)
Yes, we've got some user's feedback and I would like to report some small fixes related to Python Console and IPython. I'll create a separate pull request.

**Pull request closed without merging**